source/Timer.cpp

```diff
@@ -0,0 +1,322 @@
+			if(it != resets.end())
+				out.Write("reset", it->second);
+		}
+		if(proximity > 0.)
+		{
+			string distance = "close";
+			if(!closeTo)
+				distance = "far";
+			if(proximityCenter != nullptr)
+				out.Write("proximity", proximityCenter->Name());
+			else if(!proximityCenters.IsEmpty())
+			{
+				out.BeginChild();
+				{
+					proximityCenters.Save(out);
+				}
+				out.EndChild();
+			}
+			out.Write("proximity settings", proximity, distance);
+		}
+
+		action.Save(out);
+		resetAction.Save(out);
+
+	}
+	out.EndChild();
+}
+
+
+
+// Calculate the total time to wait, including any random value,
+// and instantiate the triggered action
+Timer Timer::Instantiate(map<string, string> &subs,
+						const System *origin, int jumps, int64_t payload) const
+{
+	Timer result;
+	result.name = name;
+	result.base = base;
+	result.rand = rand;
+	result.requireIdle = requireIdle;
+	result.requireUncloaked = requireUncloaked;
+	result.system = system;
+	result.systems = systems;
+	result.proximity = proximity;
+	result.proximityCenter = proximityCenter;
+	result.proximityCenters = proximityCenters;
+	result.closeTo = closeTo;
+	result.resetCondition = resetCondition;
+	result.repeatReset = repeatReset;
+	result.resetFired = resetFired;
+
+	result.timeElapsed = timeElapsed;
+	result.isComplete = isComplete;
+	result.isActive = isActive;
+
+	result.action = action.Instantiate(subs, origin, jumps, payload);
+	result.resetAction = resetAction.Instantiate(subs, origin, jumps, payload);
+
+	if(rand >= 1)
+		result.timeToWait = base + Random::Int(rand);
+	else
+		result.timeToWait = base;
+
+	return result;
+}
+
+
+
+bool Timer::IsComplete() const
+{
+	return isComplete;
+}
+
+
+
+bool Timer::IsOptional() const
+{
+	return optional;
+}
+
+
+
+void Timer::ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui)
+{
+	bool reset = cond == resetCondition;
+	reset |= (cond == Timer::ResetCondition::LEAVE_ZONE && resetCondition == Timer::ResetCondition::PAUSE);
+	reset |= (cond == Timer::ResetCondition::LEAVE_SYSTEM && (resetCondition == Timer::ResetCondition::PAUSE
+				|| resetCondition == Timer::ResetCondition::LEAVE_ZONE));
+	if(isActive && reset)
+	{
+		timeElapsed = 0;
+		timeToWait = base + Random::Int(rand);
+		if(repeatReset || !resetFired)
+		{
+			resetAction.Do(player, ui, mission);
+			resetFired = true;
+		}
+		isActive = false;
+	}
+}
+
+
+
+void Timer::Step(PlayerInfo &player, UI *ui)
+{
+	if(isComplete)
+		return;
+	if(!player.Flagship())
+		return;
+	if((system && player.Flagship()->GetSystem() != system) ||
+		(!systems.IsEmpty() && !systems.Matches(player.Flagship()->GetSystem())))
+	{
+		ResetOn(Timer::ResetCondition::LEAVE_SYSTEM, player, ui);
+		return;
+	}
+	if(requireIdle)
+	{
+		bool shipIdle = (!player.Flagship()->IsThrusting() && !player.Flagship()->IsSteering()
+						&& !player.Flagship()->IsReversing());
+		for(const Hardpoint &weapon : player.Flagship()->Weapons())
+			shipIdle &= !weapon.WasFiring();
```

Comment:
This isn't exactly idle in the same sense that idle is currently used with the timer ship. This isn't checking the player's speed, meaning the player could be whizzing away from the system center but not taking any actions and the timer would still tick down. Check that flagship.Velocity().LengthSquared() is below some threshold^2 (squaring is cheaper than square rooting, which Point.Length() does), potentially defined by the idle keyword. The existence of the idle keyword should still set requireIdle, but an optional value of idle could define the velocity you're supposed to be below, defaulted to some low value if it isn't present.

Reply:
Makes sense. Just to check, would it be generally considered more desirable if "idle" also meant "not firing"...?

Reply:
I think it'd make sense to make that an option.

Reply:
OK. I've added a separate node for that.